utils/math-utils/src/main/scala/pl/touk/nussknacker/engine/util/MathUtils.scala

```diff
@@ -105,17 +190,16 @@ trait MathUtils {
     SpringNumberUtils.convertNumberToTargetClass(n, promotedClass)
   }
 
-  protected trait SameNumericTypeHandler {
-    def onBytes(n1: Byte, n2: Byte): Byte
-    def onShorts(n1: Short, n2: Short): Short
-    def onInts(n1: Int, n2: Int): Int
-    def onLongs(n1: Long, n2: Long): Long
-    def onBigIntegers(n1: java.math.BigInteger, n2: java.math.BigInteger): java.math.BigInteger
-    def onFloats(n1: Float, n2: Float): Float
-    def onDoubles(n1: Double, n2: Double): Double
-    def onBigDecimals(n1: java.math.BigDecimal, n2: java.math.BigDecimal): java.math.BigDecimal
+  protected trait SameNumericTypeHandler[R] {
```

Comment:
why this R is need?

Reply:
We want to handle any possible return type without using asInstanceOf. In most cases R is still Number, but in compare it must be Int.

Reply:
Concrete return type for each method had advantage that someone won't introduce e.g. overflow bug - types will protect you from doing sth like xxx.toLong everywhere. All math operations should promote to wider type before doing operation that gives wider return type. IMO we don't need to handle compare in the same way. IMO you can assume that all numbers are Comparable[_] as well, so the only thing that you should do is to cast to Comparable and invoke compare method on them